APITests/utils.py

```diff
@@ -61,294 +62,246 @@ def format(self, record):
 MultiRow = List[Row]
 
 
-def fetch(url: str, params: dict, headers: dict = None) -> Response:
-    """
-    Trigger a get request
-    Args:
-        url (str): the url to run a given api request
-        params (dict): parameter of running a given api request
-        headers (dict): headers used for API requests. For example, authorization headers.
-    Returns:
-        Response: a response object
-    """
-    response = requests.get(url, params=params, headers=headers)
-    return response
-
-
-def send_manifest(
-    url: str, params: dict, headers: dict = None, manifest_path=None
-) -> Response:
-    """Send an API request to an endpoint
-    Args:
-        url (str): the url to run a given api request
-        params (dict): parameters of running the post request
-        headers (dict): headers used for API requests. For example, authorization headers.
-        manifest_path (str): file path of a manifest
-    Returns:
-        Response: a response object
-    """
-    wd = os.getcwd()
-    test_manifest_path = os.path.join(wd, manifest_path)
-
-    if not os.path.exists(test_manifest_path):
-        logger.error(
-            "the manifest does not exist. Please provide a valid manifest file path"
-        )
-
-    return requests.post(
-        url,
-        params=params,
-        headers=headers,
-        files={"file_name": open(test_manifest_path, "rb")},
-    )
-
+@dataclass
+class CalculateRunTime:
```

Comment:
I feel like the way that you split up this functionality works, but is more confusing than any benefit you gain out of grouping functionality like this. If I understand the logic here we can break this down into a few large "Things" that you need to accomplish:

1. You need to make the actual call to `manifest_to_send_func` to execute the code that you want to test
2. You need to collect stats on the HTTP response codes from each function call
3. You need to collect the total time it took to execute the requests as a whole.


My 2 cents after looking at this code:

1. FormatPerformanceOutput doesn't need to be a dataclass. There is nothing stateful contained within a specific instance of the class. All of the methods contained within can be individual functions.
2. https://www.geeksforgeeks.org/time-perf_counter-function-in-python/ is nicer to use than directly using time.time() to find a difference between 2 points in your code
3. A different flow of responsibility in the code could be as follows:

- You execute all of the requests inside the `with ThreadPoolExecutor() as executor:`
- From that function where you execute all requests return back a list of Responses
- Pass that list of Responses to your formatter function that converts the data from a Response into your expected strings.

This is a slight shift in code responsibility that might make some cleaner interfaces. The idea here is that each functionality interface is responsible for 1 thing, and 1 thing only.

Reply:
Thank you so much for your comment Bryan! I agree that functions within `FormatPerformanceOutput` class can be their own individual functions. But I am curious what you meant by "execute all of the requests inside the with ThreadPoolExecutor() as executor and return a list of responses"?  Currently, the function waits for all the futures to finish and return status code and execution time and so on. How's the new flow that you proposed different from the current one?

Reply:
This is not a complete script. But this is what I was thinking:
```
from time import perf_counter

# All of this code 'AAAA' would go into 1 function responsible for executing the requests
# It should return a list of HTTP responses
# AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
def execute_requests(some_arg):
    http_responses = []
    with ThreadPoolExecutor() as executor:
        # Step 1:
        # Execute all futures in the executor. Each executor should return back an HTTP Response
        # .
        # . Code to create the futures and submit them to the executor
        # .
        
        # Step 2: 
        # As futures are completed, get the HTTP Response and store it in a list
        for f in concurrent.futures.as_completed(futures):
            try:
                http_responses.append(f.result())
            except InvalidSchema as ex:
                # Handle this exception
                print(f"An exception occurred: {ex}")

# AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA

# All of this code 'BBBB' would go into another function responsible for processing the responses
# It should return the mapping of HTTP response codes
# BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB
def parse_request_responses(http_responses):
    all_status_code = {"200": 0, "500": 0, "503": 0, "504": 0}
    for response in http_responses:
        status_code = response.status_code
        if status_code in all_status_code:
            all_status_code[status_code] += 1
        else:
            all_status_code[status_code] = 1

# BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB
def run_my_test(some_arg):
    before = perf_counter()
    http_responses = execute_requests(some_arg)
    all_status_code = parse_request_responses(http_responses)
    time_diff = round(perf_counter() - before, 2)
    return time_diff, all_status_code # ect....
```

What I am suggesting is that each function does a specific task.
A: Execute the requests and return the results - This is complex enough by itself, so let's contain "how" it does it within it's own function.
B: Parse the HTTP responses and produce our expected output.

Reply:
Thanks for the example Bryan! This makes sense to me. But I am also curious if I could put `run_my_test` as C rather than group that into B because it seems like that `run_my_test` is the "final" interface that other functions in the code needs to call to execute and format HTTP responses.

Reply:
> But I am also curious if I could put run_my_test as C rather than group that into B

Yes this is my suggestion. The way I am thinking about it is that each of the APITests are setting up what the input to the tests are and then this `run_my_test` is a more generic interface that all of your APITests call into. The idea being that the only thing that needs to change test to test is the inputs. That way you are standarizing how each test is ran, and the formatted output from each of those tests.

Reply:
https://github.com/Sage-Bionetworks/synapsePythonClient/blob/develop/docs/scripts/uploadBenchmark.py#L393

Is not exactly the same thing, but it might be useful to see a slighty different application of thse ideas. I hastily wrote the code so there are some things that could be improved here.

Reply:
Thanks for your recommendation and linking an example @BryanFauble !